Web Inspector: Elements Tab: add a navigation item to control what badges are shown

dcrousso opened the pull request:
#### 5c3f6f4a6d39f6ff0fd12999c8ba73abe94d0124
<pre>
Web Inspector: Elements Tab: add a navigation item to control what badges are shown
<a href="https://bugs.webkit.org/show_bug.cgi?id=243858">https://bugs.webkit.org/show_bug.cgi?id=243858</a>

Reviewed by Patrick Angle.

This will allow us to add more badges (e.g. &quot;Scroll&quot;, &quot;Events&quot;, etc.) without overloading the DOM
tree with too many badges, as developers can turn off ones they do not find useful.

* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView):
(WI.DOMTreeContentView.prototype.get navigationItems):
(WI.DOMTreeContentView.prototype.attached):
(WI.DOMTreeContentView.prototype.detached):
(WI.DOMTreeContentView.prototype._populateConfigureDOMTreeBadgesNavigationItemContextMenu): Added.
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.css:
(.item.button.text-only.configure-dom-tree-badges): Added.
(.item.button.text-only.configure-dom-tree-badges:hover): Added.
The &quot;Badges&quot; navigation item will show a contextmenu containing checkbox items for each type of DOM
tree badge. If any badges are selected, the navigation item is considered activated. Currently there
are only two badges: &quot;flex&quot; and &quot;grid&quot;.

* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement):
(WI.DOMTreeElement.prototype.onattach):
(WI.DOMTreeElement.prototype.ondetach):
(WI.DOMTreeElement.prototype.insertChildElement):
(WI.DOMTreeElement.prototype.updateTitle):
(WI.DOMTreeElement.prototype._createBadge): Added.
(WI.DOMTreeElement.prototype._createBadges): Added.
(WI.DOMTreeElement.prototype._updateBadges): Added.
(WI.DOMTreeElement.prototype._handleLayoutFlagsChanged):
(WI.DOMTreeElement.prototype._handleShownDOMTreeBadgesChanged): Added.
(WI.DOMTreeElement.prototype._updateLayoutBadge): Deleted.
(WI.DOMTreeElement.prototype._updateLayoutBadgeStatus): Deleted.
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.css:
(.tree-outline.dom .badge): Renamed from `.tree-outline.dom .layout-badge`.
(.tree-outline.dom .badge.activated): Renamed from `.tree-outline.dom .layout-badge.activated`.
(body:not(.window-inactive, .window-docked-inactive) .tree-outline.dom:focus-within li.selected .badge): Renamed from `body:not(.window-inactive, .window-docked-inactive) .tree-outline.dom:focus-within li.selected .layout-badge`.
(@media (prefers-color-scheme: dark) .tree-outline.dom .badge): Renamed from `@media (prefers-color-scheme: dark) .tree-outline.dom .layout-badge`.
Rework how badges are created to first check that the desired badge is allowed based on the current
selections from the above navigation item. Also don&apos;t assume that there is only one badge.

* Source/WebInspectorUI/UserInterface/Views/DOMTreeOutline.js:
(WI.DOMTreeOutline.prototype.update):
Add a `showBadges` constructor option that controls whether any child `WI.DOMTreeElement` will even
try to create a badge. This ensures that badges are only created in the DOM tree in the Elements Tab.

* Source/WebInspectorUI/UserInterface/Views/HierarchicalPathNavigationItem.js:
(WI.HierarchicalPathNavigationItem.prototype.update):
Drive-by: Use `.width` instead of `.element.realOffsetWidth` as the former includes any CSS margins.

* Source/WebInspectorUI/UserInterface/Views/ButtonNavigationItem.js:
(WI.ButtonNavigationItem.prototype.get totalMargin):
(WI.ButtonNavigationItem.prototype.get textMargin): Added.
* Source/WebInspectorUI/UserInterface/Views/ButtonNavigationItem.css:
(.navigation-bar .item.button.text-only):
* Source/WebInspectorUI/UserInterface/Views/TextToggleButtonNavigationItem.js:
(WI.TextToggleButtonNavigationItem.prototype.get textMargin): Added.
* Source/WebInspectorUI/UserInterface/Views/TextToggleButtonNavigationItem.css:
(.navigation-bar .item.text-toggle.button.text-only):
(.navigation-bar .item.text-toggle.button.text-only:not(:hover, .selected)): Added.
Drive-by: Decrease text toggle `margin` and `padding` to better match text button.

* Source/WebInspectorUI/UserInterface/Views/TimelineOverview.js:
(WI.TimelineOverview):
* Source/WebInspectorUI/UserInterface/Views/TimelineOverview.css:
(.navigation-bar.timelines .item.text.enabled-timelines): Added.
* Source/WebInspectorUI/UserInterface/Views/AuditNavigationSidebarPanel.js:
(WI.AuditNavigationSidebarPanel.prototype.showDefaultContentView):
(WI.AuditNavigationSidebarPanel.prototype.initialLayout):
* Source/WebInspectorUI/UserInterface/Views/AuditNavigationSidebarPanel.css:
(.content-view.audit .message-text-view .navigation-item-help:is(.start-editing-audits, .stop-editing-audits) .navigation-bar): Deleted.
* Source/WebInspectorUI/UserInterface/Views/Main.css:
(.navigation-item-help &gt; .navigation-bar):
Drive-by: Add an icon for all other text buttons.

* Source/WebInspectorUI/UserInterface/Views/TextNavigationItem.css:
(.navigation-bar .item:first-child.text): Added.
(.navigation-bar .item:last-child.text): Added.
Drive-by: Don&apos;t add so much padding if the text is at the beginning/end of the `WI.NavigationBar`.

* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(Array.prototype.toggleIncludes):
* LayoutTests/inspector/unit-tests/array-utilities.html:
* LayoutTests/inspector/unit-tests/array-utilities-expected.txt:
Drive-by: Allow not passing a `force`, which behaves like a flip-flop.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

Canonical link: <a href="https://commits.webkit.org/253579@main">https://commits.webkit.org/253579@main</a>
</pre>






















<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ddef6f48509c021a045c35fb92b3a203a858116

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86459 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30380 "Built successfully") | [✅ ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/17426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95304 "Built successfully") | [✅ ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/149015 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/28784 "Built successfully") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/78613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90548 "failed Failed to checkout and rebase branch from PR 3251") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92075 "Passed tests") | [✅ ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/78613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/17426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/78613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26691 "Built successfully") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/17426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26605 "Built successfully") | [✅ ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/17426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28284 "Built successfully") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/985 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28225 "Built successfully") | [✅ ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/17426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -51,7 +51,9 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
-        this._layoutBadgeElement = null;
+
+        this._showBadges = !!showBadges;
+        this._elementForBadge = new Map;
```
The singular variable name makes me think this is the element in which the badge is placed, not that it is a map of badge type to element. Maybe `_badgeElementsByType` or something similar?

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2074,29 +2103,38 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
-            this._layoutBadgeElement.removeAttribute("style");
+    _updateBadges()
+    {
+        for (let [badge, badgeElement] of this._elementForBadge) {
```
I think `[badgeType, badgeElement]` would make more sense, since it is easy to think of the "badge" as the thing that is actually shown on the screen.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2074,29 +2103,38 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
+        for (let [badge, badgeElement] of this._elementForBadge) {
+            switch (badge) {
+            case WI.DOMNode.LayoutFlag.Grid:
+            case WI.DOMNode.LayoutFlag.Flex: {
```
Should these be `WI.DOMTreeElement.Badge.*` values to match the types used to insert the badge elements into the map? (Yes, they have the same values, but probably still better to be consistent.)

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2122,6 +2160,12 @@ WI.DOMTreeElement.BreakpointStatus = {
+    Flex: "flex",
+    Grid: "grid",
+};
+WI.settings.shownDOMTreeBadges = new WI.Setting("shown-dom-tree-badges", [WI.DOMTreeElement.Badge.Flex, WI.DOMTreeElement.Badge.Grid]);
```
`shown` feels like the wrong adjective here. It make me think that it means actually shown to the user, which won't always be true on every page. Maybe `enabledDOMTreeBadgeTypes`?

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2122,6 +2160,12 @@ WI.DOMTreeElement.BreakpointStatus = {
     DisabledBreakpoint: Symbol("disabled-breakpoint"),
 };
 
+WI.DOMTreeElement.Badge = {
```
Same argument as above, `BadgeType`?

patrickangle reviewed (commented):
Also, as discussed offline I'd like to see a less prominent style for the button in the bar.

dcrousso committed:
Web Inspector: Elements Tab: add a navigation item to control what badges are shown
https://bugs.webkit.org/show_bug.cgi?id=243858

Reviewed by Patrick Angle.

This will allow us to add more badges (e.g. "Scroll", "Events", etc.) without overloading the DOM
tree with too many badges, as developers can turn off ones they do not find useful.

* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView):
(WI.DOMTreeContentView.prototype.get navigationItems):
(WI.DOMTreeContentView.prototype.attached):
(WI.DOMTreeContentView.prototype.detached):
(WI.DOMTreeContentView.prototype._populateConfigureDOMTreeBadgesNavigationItemContextMenu): Added.
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.css:
(.item.button.text-only.configure-dom-tree-badges): Added.
(.item.button.text-only.configure-dom-tree-badges:hover): Added.
The "Badges" navigation item will show a contextmenu containing checkbox items for each type of DOM
tree badge. If any badges are selected, the navigation item is considered activated. Currently there
are only two badges: "flex" and "grid".

* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement):
(WI.DOMTreeElement.prototype.onattach):
(WI.DOMTreeElement.prototype.ondetach):
(WI.DOMTreeElement.prototype.insertChildElement):
(WI.DOMTreeElement.prototype.updateTitle):
(WI.DOMTreeElement.prototype._createBadge): Added.
(WI.DOMTreeElement.prototype._createBadges): Added.
(WI.DOMTreeElement.prototype._updateBadges): Added.
(WI.DOMTreeElement.prototype._handleLayoutFlagsChanged):
(WI.DOMTreeElement.prototype._handleShownDOMTreeBadgesChanged): Added.
(WI.DOMTreeElement.prototype._updateLayoutBadge): Deleted.
(WI.DOMTreeElement.prototype._updateLayoutBadgeStatus): Deleted.
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.css:
(.tree-outline.dom .badge): Renamed from `.tree-outline.dom .layout-badge`.
(.tree-outline.dom .badge.activated): Renamed from `.tree-outline.dom .layout-badge.activated`.
(body:not(.window-inactive, .window-docked-inactive) .tree-outline.dom:focus-within li.selected .badge): Renamed from `body:not(.window-inactive, .window-docked-inactive) .tree-outline.dom:focus-within li.selected .layout-badge`.
(@media (prefers-color-scheme: dark) .tree-outline.dom .badge): Renamed from `@media (prefers-color-scheme: dark) .tree-outline.dom .layout-badge`.
Rework how badges are created to first check that the desired badge is allowed based on the current
selections from the above navigation item. Also don't assume that there is only one badge.

* Source/WebInspectorUI/UserInterface/Views/DOMTreeOutline.js:
(WI.DOMTreeOutline.prototype.update):
Add a `showBadges` constructor option that controls whether any child `WI.DOMTreeElement` will even
try to create a badge. This ensures that badges are only created in the DOM tree in the Elements Tab.

* Source/WebInspectorUI/UserInterface/Views/HierarchicalPathNavigationItem.js:
(WI.HierarchicalPathNavigationItem.prototype.update):
Drive-by: Use `.width` instead of `.element.realOffsetWidth` as the former includes any CSS margins.

* Source/WebInspectorUI/UserInterface/Views/ButtonNavigationItem.js:
(WI.ButtonNavigationItem.prototype.get totalMargin):
(WI.ButtonNavigationItem.prototype.get textMargin): Added.
* Source/WebInspectorUI/UserInterface/Views/ButtonNavigationItem.css:
(.navigation-bar .item.button.text-only):
* Source/WebInspectorUI/UserInterface/Views/TextToggleButtonNavigationItem.js:
(WI.TextToggleButtonNavigationItem.prototype.get textMargin): Added.
* Source/WebInspectorUI/UserInterface/Views/TextToggleButtonNavigationItem.css:
(.navigation-bar .item.text-toggle.button.text-only):
(.navigation-bar .item.text-toggle.button.text-only:not(:hover, .selected)): Added.
Drive-by: Decrease text toggle `margin` and `padding` to better match text button.

* Source/WebInspectorUI/UserInterface/Views/TimelineOverview.js:
(WI.TimelineOverview):
* Source/WebInspectorUI/UserInterface/Views/TimelineOverview.css:
(.navigation-bar.timelines .item.text.enabled-timelines): Added.
* Source/WebInspectorUI/UserInterface/Views/AuditNavigationSidebarPanel.js:
(WI.AuditNavigationSidebarPanel.prototype.showDefaultContentView):
(WI.AuditNavigationSidebarPanel.prototype.initialLayout):
* Source/WebInspectorUI/UserInterface/Views/AuditNavigationSidebarPanel.css:
(.content-view.audit .message-text-view .navigation-item-help:is(.start-editing-audits, .stop-editing-audits) .navigation-bar): Deleted.
* Source/WebInspectorUI/UserInterface/Views/Main.css:
(.navigation-item-help > .navigation-bar):
Drive-by: Add an icon for all other text buttons.

* Source/WebInspectorUI/UserInterface/Views/TextNavigationItem.css:
(.navigation-bar .item:first-child.text): Added.
(.navigation-bar .item:last-child.text): Added.
Drive-by: Don't add so much padding if the text is at the beginning/end of the `WI.NavigationBar`.

* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(Array.prototype.toggleIncludes):
* LayoutTests/inspector/unit-tests/array-utilities.html:
* LayoutTests/inspector/unit-tests/array-utilities-expected.txt:
Drive-by: Allow not passing a `force`, which behaves like a flip-flop.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

Canonical link: https://commits.webkit.org/253579@main
